src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,156 @@
+        if (size == 0) {
+            firstNode = newNode;
+            lastNode = newNode;
+        } else {
+            lastNode.nextElement = newNode;
+            newNode.prevElement = lastNode;
+            lastNode = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> currentNode = findByIndex(index);
+            Node<T> newNode = new Node<>(currentNode.prevElement, value, currentNode);
+            newNode.nextElement.prevElement = newNode;
+            if (index != 0) {
+                newNode.prevElement.nextElement = newNode;
+            } else {
+                firstNode = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T current : list) {
+            add(current);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index > size - 1) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        Node<T> currentNode = findByIndex(index);
+        return currentNode.currentElement;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index > size - 1) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        Node<T> currentNode = findByIndex(index);
+        T deletedValue = currentNode.currentElement;
+        currentNode.currentElement = value;
+        return deletedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index > size - 1) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        Node<T> currentNode = findByIndex(index);
+        deleteLinks(currentNode);
+        return currentNode.currentElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = firstNode;
+        while (currentNode != null) {
+            if (currentNode.currentElement == null && object == null
+                    || currentNode != null && currentNode.currentElement.equals(object)) {
+                deleteLinks(currentNode);
+                return true;
+            }
+            currentNode = currentNode.nextElement;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findByIndex(int index) {
+        int counter = 0;
+        Node<T> currentNode = firstNode;
+        while (currentNode.nextElement != null) {
```

Comment:
perhaps better use loop: for...

Reply:
but as far as i know, for linked list better use while than for each